src/backend/controllers/marketController.js

```diff
@@ -0,0 +1,308 @@
+}
+
+// Transaction between players
+exports.marketTransaction = async (req: $Request, res: $Response) => {
+    // Validation
+    const errors = validationResult(req);
+
+    if (!errors.isEmpty()) {
+        // Return 400 for a bad request
+        return res
+            .status(400)
+            .json({ errors: errors.array() });
+    }
+
+    // Deconstruct request body
+    const { buyerId, sellerId, saleId } = req.body;
+
+    // Check if users exist
+    let buyer = await User.findById(buyerId);
+    const seller = await User.findById(sellerId);
+    if (!buyer) {
+        return res  
+            .status(400)
+            .json({ msg: 'Buyer user does not exist' })
+    }
+
+    if (!seller) {
+        return res  
+            .status(400)
+            .json({ msg: 'Seller user does not exist' })
+    }
+
+    // Check if sale exists
+    const sale = await MarketSale.findById(saleId);
+    if (!sale) {
+        return res
+            .status(400)
+            .json({ msg: 'Sale post does not exist' })
+    }
+
+    // Check if buyer has enough money
+    if (buyer.money < sale.salePrice) {
+        return res
+            .status(400)
+            .json({ msg: 'Not enough money for this transaction.' })
+    }
+
+    // Buying a Tank or a Component
+    // Note: Depending on the itemType, if it is a tank,
+    // itemID is the ObjectId of the tank, otherwise it is
+    // the string identifier of the component/Casus block equivalent
+    // to the property in the User Model.
+    if (sale.itemType === 'tank') {
+        try {         
+            // Check if tank exists
+            const tank = await Tank.findById(sale.itemId);
+            if (!tank) {
+                return res
+                    .status(400)
+                    .json({ msg: 'Tank does not exist' })
+            }
+
+            // Start transaction
+            buyer = await User.findByIdAndUpdate(buyerId, { $inc: { money: (sale.salePrice * -1) } }, { new: true });
+            await User.findByIdAndUpdate(sellerId, { $inc: { money: sale.salePrice } });
+            await Tank.findByIdAndUpdate(sale.itemId, { $set: { userId: buyerId } });
+
+            // Remove the sale from database
+            await MarketSale.deleteOne({ id: saleId });
+
+            // Return current buyer
+            res.status(201).json(buyer);
+
+        } catch (err) {
+            console.error(err.message);
+            res.status(500).json({ msg: 'Cannot make Market Transaction.' });
+        }
+    } else { 
+        // Component 
+        if (sale.itemType === 'component') {       
+            try {
+                // Start transaction
+                buyer = await User.findByIdAndUpdate(buyerId, { $inc: { money: (sale.salePrice * -1) } }, { new: true });
+                await User.findByIdAndUpdate(sellerId, { $inc: { money: sale.salePrice } });
+                // Add items to buyer
+                buyer['inventory']['tankComponents'][sale.itemId] += sale.amount;
```

Comment:
Is this a problem if sale.itemId is invalid?

Reply:
Yeah good catch. I have to just add a check in the addMarketSale function to ensure that field exists = the itemId is a good string to use. Once a sale is stored with valid properties, unless we go total mean mode and actually change the field in the database manually, we can be assured that the itemId is valid coming into here. Note that this check only needs to apply for the tankComponents and casusBlocks, since tanks can be easily verified by their ObjectIDs